### ISSUE TITLE
Immutables with Retrofit 2

### ISSUE THREAD

**Issue by devilmac** [immutables/immutables#287]
Hi,
I added this great library to my Android application, which use Retrofit 2 to make HTTP requests and Gson to deserialize the responses.
I'm having a problem using response classes as immutable, because Retrofit is unable to convert JSON response to the specific response class without the specific type adapter.
I added TypeAdapters inside my service adapter class, following instruction described here http://immutables.github.io/json.html#type-adapter-registration:

```java
        GsonBuilder gsonBuilder = new GsonBuilder();
        for (TypeAdapterFactory factory : ServiceLoader.load(TypeAdapterFactory.class)) {
            gsonBuilder.registerTypeAdapterFactory(factory);
        }
```

but it seems don't work; the error I receive is as follow:

```
at com.google.gson.internal.UnsafeAllocator$1.newInstance(UnsafeAllocator.java:48)
03-19 14:09:40.104 963-1027/? A/art: art/runtime/java_vm_ext.cc:410]   at com.google.gson.internal.ConstructorConstructor$14.construct(ConstructorConstructor.java:223)
03-19 14:09:40.104 963-1027/? A/art: art/runtime/java_vm_ext.cc:410]   at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:206)
03-19 14:09:40.104 963-1027/? A/art: art/runtime/java_vm_ext.cc:410]   at retrofit2.converter.gson.GsonResponseBodyConverter.convert(GsonResponseBodyConverter.java:37)
03-19 14:09:40.104 963-1027/? A/art: art/runtime/java_vm_ext.cc:410]   at retrofit2.converter.gson.GsonResponseBodyConverter.convert(GsonResponseBodyConverter.java:25)
03-19 14:09:40.104 963-1027/? A/art: art/runtime/java_vm_ext.cc:410]   at retrofit2.ServiceMethod.toResponse(ServiceMethod.java:116)
03-19 14:09:40.104 963-1027/? A/art: art/runtime/java_vm_ext.cc:410]   at retrofit2.OkHttpCall.parseResponse(OkHttpCall.java:211)
03-19 14:09:40.104 963-1027/? A/art: art/runtime/java_vm_ext.cc:410]   at retrofit2.OkHttpCall.execute(OkHttpCall.java:174)
```

I'm doing something wrong? Can someone helps me?
Thanks

**Comment by elucash** [comment]
Hi Federico
I just will be unable to actually check it if it works right away, but I'll try to lay out all the steps I see necessary for this to work:

1) Make sure you've created/generated Gson Type adapters. It could be done with some variations, but I would suggest to create a `package-info.java` and put `Gson.TypeAdapter` annotation on a package, this will generate `TypeAdapterFactory` for the whole package making it easier to register and have a compact code overall.

```java
// com/acme/jsontypes/package-info.java
@org.immutables.gson.Gson.TypeAdapters
package com.acme.jsontypes;
```

```java
// com/acme/jsontypes/SampleType.java
package com.acme.jsontypes;
import org.immutables.value.Value;
// Just a sample type for JSON serialization
@Value.Immutable interface SampleType {
   int getId();
   String getName();
}
```

When compiling with the `Immutables` annotation processor, `ImmutableSampleType` and `GsonAdaptersJsontypes` classes will be also generated, where `GsonAdaptersJsontypes` is a type adapter factory derived for the all immutable types compiled in a package. Note that when adding new or changing immutable types in the package, you will need to fully recompile package (or whole module) so all classes will be included (or you can place `Gson.TypeAdapters` annotation on each immutable type, this will make every type have it's own adapter factory, which is more reliable during incremental compilation but may be more tedious to register).

2) Add custom factories / adapters to Gson. As it was pointed out here #230 , `ServiceLoader` is not a best way to register adapters on Android.

```java
com.google.gson.Gson gson = new com.google.gson.GsonBuilder()
      .registerTypeAdapterFactory(new GsonAdaptersJsontypes())
      .create();
```

3) The last step before using it would be to register `Gson` with the `Retrofit`. Following documentation, you need to use `converter-gson` module, you probably already using it. The modified example for the documenation would look like

```java
Retrofit retrofit = new Retrofit.Builder()
    .baseUrl("https://api.github.com")
    .addConverterFactory(GsonConverterFactory.create(gson)) // <-- here we pass gson instance
    .build();

GitHubService service = retrofit.create(GitHubService.class);
```

4) Now everything should be setup to use properly adapter `SampleType` and other registered immutable classes with retrofit. The fact that you had error inside `ReflectiveTypeAdapterFactory` indicated that type adapters was either not generated (first step) or not registered properly (second or third step).
You can still use `ServiceLoader` to easy register all adapters from the classpath, but it may be slow; and it uses `/META-INF/services/...` files generated, which may not propagate properly during you builder, for example apk might not have them for some reason, depending on Grails or it's plugins configurations etc. Having per-package factories (and registering new instances of them) seems to work best for Android apps.

Hope this helps, thank you!

**Comment by devilmac** [comment]
Hi @elucash ,
thank you for your help, it was very useful. What I missed is that I did not understand the step to configure Gson.TypeAdapter using package-info.java file and I used wrong directive on build.gradle of my network module for "org.immutables:value:2.1.14": instead of
`apt "org.immutables:value:2.1.14" // for annotation processor
  provided "org.immutables:value:2.1.14" // for annotations
  provided "org.immutables:builder:2.1.14" // for annotations
  provided "org.immutables:gson:2.1.14"`
I was using
`compile "org.immutables:value:2.1.14"
 compile "org.immutables:gson:2.1.14"`
Following your steps I was able to get it work.

Thanks a lot again.
